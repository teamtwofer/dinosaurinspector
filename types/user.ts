export interface IUser {
  id: number;
  email: string;
  name: string;
}

export interface IRegisterUser {
  email: string;
  name: string;
  password: string;
}
