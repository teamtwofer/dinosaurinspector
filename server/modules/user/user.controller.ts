import {
  Body,
  Controller,
  Delete,
  Get,
  HttpStatus,
  Post,
  Req,
  Res,
} from '@nestjs/common';
import { HttpException } from '@nestjs/core';
import { Response } from 'express';

import { IRegisterUser } from '../../../types/user';
import { UserSerializer } from './user.serializer';
import { UserService } from './user.service';

@Controller()
export class UserController {
  service: UserService;
  private serializer: UserSerializer;

  constructor(userService: UserService, userSerializer: UserSerializer) {
    this.service = userService;
    this.serializer = userSerializer;
  }

  @Get('user')
  async getUser(@Req() req, @Res() res: Response) {
    res.status(HttpStatus.OK).json(this.serializer.serializeFull(req.user));
  }

  @Post('user/token')
  async generateToken(@Res() res, @Body('user') user: Partial<IRegisterUser>) {
    try {
      const token = await this.service.generateToken(user);
      res.status(HttpStatus.OK).json({ token });
    } catch (error) {
      throw new HttpException(error.message, HttpStatus.NOT_ACCEPTABLE);
    }
  }

  @Delete('user/:id')
  async deleteUser(@Req() req, @Res() res: Response) {
    const { user } = req;
    const deletedUser = await this.service.remove(user);
    res.status(HttpStatus.OK).json(this.serializer.serializeFull(deletedUser));
  }
}
