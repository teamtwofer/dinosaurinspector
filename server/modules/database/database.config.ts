import { Component } from '@nestjs/common';
import { ConnectionOptions } from 'typeorm';
import { User } from '../../entities/user.entity';

@Component()
export class DatabaseConfig {
  getConfiguration(): ConnectionOptions {
    const { env } = process;
    const isTesting = env.NODE_ENV === 'test';
    return {
      autoSchemaSync: true,
      driver: {
        database: isTesting ? 'dinosaur_testing' : 'dinosaur_development',
        host: 'localhost',
        password: 'potato',
        port: 5432,
        type: isTesting ? 'sqlite' : 'postgres',
        username: 'bbayard',
      },
      entities: [User],
      logging: {
        logFailedQueryError: !isTesting,
        logQueries: !isTesting,
      },
      name: `${Date.now()}`,
      type: (isTesting ? 'sqlite' : 'postgres') as any, // this field isn't necessary, but TS complains without it
    };
  }
}
