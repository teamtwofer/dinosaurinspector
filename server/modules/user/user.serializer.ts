import { Component } from '@nestjs/common';
import { IUser } from '../../../types/user';
import { User } from '../../entities/user.entity';

@Component()
export class UserSerializer {
  serializeFull({ email, id, name, createdAt, updatedAt }: User): IUser {
    return { email, id, name, createdAt, updatedAt: updatedAt || null };
  }

  serializeTokenAndUser(user: User, token: string) {
    return { token, user: this.serializeFull(user) };
  }
}
