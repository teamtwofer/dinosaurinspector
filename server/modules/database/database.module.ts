import { Module } from '@nestjs/common';
import { DatabaseConfig } from './database.config';
import { DatabaseService } from './database.service';

@Module({
  components: [DatabaseService, DatabaseConfig],
  exports: [DatabaseService],
})
export class DatabaseModule {}
