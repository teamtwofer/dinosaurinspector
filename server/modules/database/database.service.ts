import { Component } from '@nestjs/common';
import {
  Connection,
  createConnection,
  EntityManager,
  ObjectType,
  Repository,
} from 'typeorm';

import { DatabaseConfig } from './database.config';

@Component()
export class DatabaseService {
  private _connection: Connection | null = null;

  constructor(private readonly databaseConfig: DatabaseConfig) {}

  get connection(): Promise<Connection> {
    if (this._connection) {
      return Promise.resolve(this._connection);
    }
    return createConnection(this.databaseConfig.getConfiguration())
      .then(connection => {
        console.info(this.databaseConfig.getConfiguration());
        this._connection = connection;
        return connection;
      })
      .catch(e => {
        // tslint:disable:no-console
        console.log('There was an error connecting to the database');
        console.log(e.message);
        console.info(this.databaseConfig.getConfiguration());
        // tslint:enable:no-console
        throw e;
      });
  }

  async getEntityManager(): Promise<EntityManager> {
    return (await this.connection).manager;
  }

  async getRepository<T>(
    entityClassOrName: ObjectType<T> | string
  ): Promise<Repository<T>> {
    return (await this.connection).getRepository<T>(entityClassOrName);
  }
}
