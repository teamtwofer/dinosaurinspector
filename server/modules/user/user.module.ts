import { MiddlewaresConsumer, Module, RequestMethod } from '@nestjs/common';
import { AuthMiddleware } from '../auth/auth.middleware';
import { DatabaseConfig } from '../database/database.config';
import { DatabaseModule } from '../database/database.module';
import { DatabaseService } from '../database/database.service';
import { EmailModule } from '../email/email.module';
import { EmailService } from '../email/email.service';
import { ForgotPasswordModule } from '../forgot-password/forgot-password.module';
import { ForgotPasswordService } from '../forgot-password/forgot-password.service';
import { UserController } from './user.controller';
import { UserSerializer } from './user.serializer';
import { UserService } from './user.service';

@Module({
  components: [
    UserService,
    DatabaseConfig,
    DatabaseService,
    UserSerializer,
    ForgotPasswordService,
    EmailService,
  ],
  controllers: [UserController],
  modules: [DatabaseModule, ForgotPasswordModule, EmailModule],
})
export class UserModule {
  configure(consumer: MiddlewaresConsumer) {
    consumer.apply(AuthMiddleware).forRoutes(
      {
        method: RequestMethod.GET,
        path: 'user',
      },
      {
        method: RequestMethod.DELETE,
        path: 'user/:id',
      }
    );
  }
}
