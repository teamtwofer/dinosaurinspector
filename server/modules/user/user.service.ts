import { Component } from '@nestjs/common';
import { autobind } from 'core-decorators';
import * as jwt from 'jsonwebtoken';
import { Repository } from 'typeorm';
import bcrypt = require('bcrypt');

import { lang } from '../../../lang';
import { ICrud } from '../../../types/crud';
import { IRegisterUser } from '../../../types/user';
import { User } from '../../entities/user.entity';
import key from '../../entities/user.key';
import { DatabaseService } from '../database/database.service';

@Component()
export class UserService implements ICrud<User, IRegisterUser> {
  constructor(public databaseService: DatabaseService) {
    if (process.env.NODE_ENV === 'development') {
      this.seed();
    }
  }

  private get repository(): Promise<Repository<User>> {
    return this.databaseService.getRepository(User);
  }

  async seed() {
    const repo = await this.repository;
    const count = await repo.count();
    if (count === 0) {
      const ben: IRegisterUser = {
        email: 'ben@twofer.co',
        name: 'Ben',
        password: 'password',
      };
      const users = await this.add(ben);

      return users;
    }

    return repo.find();
  }

  @autobind
  async add(userToRegister: IRegisterUser) {
    const user = await User.from(userToRegister);
    return (await this.repository).persist(user);
  }

  @autobind
  async addAll(...users: User[]) {
    return (await this.repository).persist(users);
  }

  @autobind
  async getAll() {
    return (await this.repository).find();
  }

  @autobind
  async get(id: number) {
    return (await this.repository).findOneById(id);
  }

  @autobind
  async update(user: User) {
    return (await this.repository).persist(user);
  }

  @autobind
  async remove(user: User) {
    return (await this.repository).remove(user);
  }

  @autobind
  async generateToken({
    email,
    password,
  }: Partial<IRegisterUser>): Promise<string> {
    const user = await (await this.repository).findOne({ email });
    if (!user) {
      throw new Error(lang.INVALID_EMAIL_OR_PASSWORD());
    }

    const validPassword = await bcrypt.compare(password, user.hashedPassword);

    if (!validPassword) {
      throw new Error(lang.INVALID_EMAIL_OR_PASSWORD());
    }

    const { id } = user;

    return jwt.sign({ id } as object, key, {
      expiresIn: 60 * 60 * 24 * 14,
    });
  }

  @autobind
  async validateToken(token: string): Promise<number> {
    return await new Promise<number>((res, rej) => {
      jwt.verify(token, key, (err, decoded: { id: number }) => {
        if (err) {
          return rej(err);
        }
        return res(decoded.id);
      });
    });
  }
}
