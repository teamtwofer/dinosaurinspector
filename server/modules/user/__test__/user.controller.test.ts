import 'reflect-metadata';

import { Test } from '@nestjs/testing';
import { Response } from 'express';

import { IRegisterUser } from '../../../../types/user';
import { User } from '../../../entities/user.entity';
import { DatabaseConfig } from '../../database/database.config';
import { DatabaseService } from '../../database/database.service';
import { UserController } from '../user.controller';
import { UserSerializer } from '../user.serializer';
import { UserService } from '../user.service';

describe('UserController', () => {
  let controller: UserController;
  let db: DatabaseService;
  let jsonResonse: object;
  let ben: IRegisterUser;
  let user: User;

  const response: Response = {
    status: () => ({
      json: (j: object) => {
        jsonResonse = j;
      },
    }),
  } as any;

  beforeEach(async () => {
    Test.createTestingModule({
      components: [
        UserService,
        UserSerializer,
        DatabaseConfig,
        DatabaseService,
      ],
      controllers: [UserController],
    });

    ben = {
      email: 'ben@twofer.co',
      name: 'ben',
      password: 'potato',
    };

    controller = Test.get<UserController>(UserController);
    db = Test.get(DatabaseService);
    jsonResonse = {};
    user = await controller.service.add(ben);
    ben.password = 'potato';
    return;
  });

  afterEach(async () => {
    const users = await controller.service.getAll();
    await Promise.all(users.map(controller.service.remove));
    const connection = await db.connection;
    await connection.dropDatabase();
    await connection.close();
    return;
  });

  it('GET /user/:id', async () => {
    await controller.getUser({ user }, response);
    expect(jsonResonse).toMatchSnapshot();
  });

  it('GET /user/token', async () => {
    await controller.generateToken(response, ben);
    expect((jsonResonse as any).token).toBeDefined();
  });

  it('DELETE /user/:id', async () => {
    await controller.deleteUser({ user: { ...user } }, response);
    expect(jsonResonse).toMatchSnapshot();
  });
});
