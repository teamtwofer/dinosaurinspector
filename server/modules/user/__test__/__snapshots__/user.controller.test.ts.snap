// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserController DELETE /user/:id 1`] = `
Object {
  "email": "ben@twofer.co",
  "id": undefined,
  "name": "ben",
}
`;

exports[`UserController GET /user/:id 1`] = `
Object {
  "email": "ben@twofer.co",
  "id": 1,
  "name": "ben",
}
`;
